Source/Core/Core/Debugger/Thread.h

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <array>
+#include <utility>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+
+namespace Dolphin_Debugger::Thread
+{
+constexpr u32 STACK_MAGIC = 0xDEADBABE;
+
+template <class C>
+struct Queue
+{
+  u32 head;
+  u32 tail;
+};
+template <class C>
+struct Link
+{
+  u32 next;
+  u32 prev;
+};
+
+struct OSMutex;
+struct OSThread;
+
+using OSThreadQueue = Queue<OSThread>;
+using OSThreadLink = Link<OSThread>;
+
+using OSMutexQueue = Queue<OSMutex>;
+using OSMutexLink = Link<OSMutex>;
+
+using ThreadList = std::vector<std::pair<u32, OSThread>>;
+
+struct OSContext
+{
+  enum class State : u16
+  {
+    HAS_FPU = 1,
+    HAS_EXCEPTION = 2
+  };
+  std::array<u32, 32> gpr;
+  u32 cr;
+  u32 lr;
+  u32 ctr;
+  u32 xer;
+  std::array<double, 32> fpr;
+  u64 fpscr;
+  u32 srr0;
+  u32 srr1;
+  u32 dummy;
+  State state;
+  std::array<u32, 8> gqr;
+  u32 psf_padding = 0;
+  std::array<double, 32> psf;
+};
+
+struct OSThread
+{
+  OSContext context;
+
+  u16 state;               // Thread state (ready, running, waiting, moribund)
+  u16 is_detached;         // Is thread detached
+  s32 suspend;             // Suspended if greater than zero
+  s32 effective_priority;  // Effective priority
+  s32 base_priority;       // Base priority
+  u32 exit_code_addr;      // Exit value address
+
+  u32 queue_addr;           // Address of the queue the thread is on
+  OSThreadLink queue_link;  // Used to traverse the thread queue
+  // OSSleepThread uses it to insert the current thread at the end of the thread queue
+
+  OSThreadQueue join_queue;  // Threads waiting to be joined
+
+  u32 mutex_addr;            // Mutex waiting
+  OSMutexQueue mutex_queue;  // Mutex owned
+
+  OSThreadLink thread_link;  // Link containing all active threads
+
+  // The STACK_MAGIC is written at stack_end
+  u32 stack_addr;
+  u32 stack_end;
+
+  s32 error;                    // errno value
+  std::array<u32, 2> specific;  // Pointers to data (can be used to store thread names)
+};
+
+struct OSMutex
+{
+  OSThreadQueue thread_queue;  // Threads waiting to own the mutex
+  u32 owner_addr;              // Thread owning the mutex
+  s32 lock_count;              // Mutex lock count
+  OSMutexLink link;            // Used to traverse the thread's mutex queue
+  // OSLockMutex uses it to insert the acquired mutex at the end of the queue
+};
+
+void ReadContext(u32 addr, OSContext* ctx);
+void DumpContext(OSContext* ctx);
```

Comment:
Given this is dumping the context, this parameter can probably be const qualified. Ditto for `DumpMutex` and `DumpThread`